src/btf.cpp

```diff
@@ -0,0 +1,216 @@
+#include <iostream>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <unistd.h>
+#include <sys/utsname.h>
+#include <string.h>
+#include "btf.h"
+#include "types.h"
+#include "bpftrace.h"
+
+#ifdef HAVE_LIBBPF_BTF_DUMP
+#include <linux/bpf.h>
+#include <linux/btf.h>
+#include <bpf/btf.h>
+#include <bpf/libbpf.h>
+
+namespace bpftrace {
+
+static unsigned char *get_data(const char *file, ssize_t *sizep)
+{
+  ssize_t ret;
+  int fd;
+
+  fd = open(file, O_RDONLY);
+  if (fd < 0)
+    return NULL;
+
+  struct stat st;
+
+  if (fstat(fd, &st))
+  {
+    close(fd);
+    return NULL;
+  }
+
+  unsigned char *data;
+  unsigned int size;
+
+  size = st.st_size;
+
+  data = (unsigned char *) malloc(size);
+  if (!data)
+  {
+    close(fd);
+    return NULL;
+  }
+
+  ret = read(fd, data, size);
+  if (ret <= 0)
+  {
+    close(fd);
+    free(data);
+    return NULL;
+  }
+
+  close(fd);
+
+  *sizep = size;
+  return data;
+}
+
+static struct btf* btf_raw(char *file)
+{
+  unsigned char *data;
+  ssize_t size;
+  struct btf *btf;
+
+  data = get_data(file, &size);
+  if (data == NULL)
+  {
+    std::cerr << "BTF: failed to read data from: " << file << std::endl;
+    return NULL;
+  }
+
+  btf = btf__new(data, (__u32) size);
+  free(data);
+  return btf;
+}
+
+static int libbpf_print(enum libbpf_print_level level, const char *msg, va_list ap)
+{
+  fprintf(stderr, "BTF: (%d) ", level);
+  return vfprintf(stderr, msg, ap);
+}
+
+BTF::BTF(void) : btf(NULL), state(NODATA)
+{
+  char err_buf[256];
+  bool free_path = false;
+  char *path;
+
+  // Try to get BTF file from BPFTRACE_BTF_TEST env
+  path = std::getenv("BPFTRACE_BTF_TEST");
+  if (!path)
+  {
+    struct utsname uts;
+
+    if (uname(&uts))
+    {
+      std::cerr << "BTF: failed to get uname" << std::endl;
+      return;
+    }
+
+    asprintf(&path, "/lib/modules/%s/build/vmlinux", uts.release);
```

Comment:
FYI, I'm working on a patch for the kernel that will expose kernel BTF through procfs or sysfs, so it will be a stable path, hopefully making this searching of kernel BTF more reliable.

Reply:
cool, also it'll be available without debuginfo package,
could you please CC me on that kernel change, thanks

Reply:
sure, will do